File: solutions/javascript/grade-school/1/grade-school.js
export class GradeSchool {
  #roster   = new Map();
  #students = new Set();

  roster() {
    let roster = Object.fromEntries(this.#roster);

    for (const key of this.#roster.keys()) {
      roster[key] = [...this.#roster.get(key)];
    }

    return roster;
  }

  add(name, grade) {
    if (this.#students.has(name)) {
      for (const names of this.#roster.values()) {
        const index = names.indexOf(name);
        if (index !== -1) {
          names.splice(index, 1);
          break;
        }
      }
    }

    this.#students.add(name);

    if (!this.#roster.has(grade)) {
      this.#roster.set(grade, [name]);
    }
    else {
      this.#roster.get(grade).push(name);
      this.#roster.get(grade).sort();
    }
  }

  grade(grade) {
    if (!this.#roster.has(grade)) {
      return [];
    }

    return [...this.#roster.get(grade)];
  }
}

File: solutions/csharp/grade-school/1/GradeSchool.cs
using System.Collections.Generic;
using System.Linq;

public class GradeSchool
{
    private class Student
    {
        public Student(string name, int grade)
        {
            Name  = name;
            Grade = grade;
        }

        public string Name  { get; }
        public int    Grade { get; }
    }

    private IList<Student> _students;

    public GradeSchool()
        => _students = new List<Student>();

    public void Add(string student, int grade)
        => _students.Add(new Student(student, grade));

    public IEnumerable<string> Roster()
        => _students
            .OrderBy(s => s.Grade)
            .ThenBy(s => s.Name)
            .Select(s => s.Name);

    public IEnumerable<string> Grade(int grade)
        => _students
            .Where(s => s.Grade == grade)
            .OrderBy(s => s.Name)
            .Select(s => s.Name);
}